wen-restart/src/wen_restart.rs

```diff
@@ -1,58 +1,210 @@
+    let mut last_voted_fork_slots_aggregate = LastVotedForkSlotsAggregate::new(
+        root_slot,
+        REPAIR_THRESHOLD,
+        root_bank.epoch_stakes(root_bank.epoch()).unwrap(),
+        &progress
+            .my_last_voted_fork_slots
+            .as_ref()
+            .unwrap()
+            .last_vote_fork_slots,
+        &cluster_info.id(),
+    );
+    let mut cursor = solana_gossip::crds::Cursor::default();
+    let mut is_full_slots = HashSet::new();
+    progress.last_voted_fork_slots_aggregate = Some(LastVotedForkSlotsAggregateRecord {
+        received: HashMap::new(),
+    });
+    loop {
+        let start = timestamp();
+        let new_last_voted_fork_slots = cluster_info.get_restart_last_voted_fork_slots(&mut cursor);
+        let result = last_voted_fork_slots_aggregate.aggregate(
+            new_last_voted_fork_slots,
+            progress.last_voted_fork_slots_aggregate.as_mut().unwrap(),
+        );
+        let mut filtered_slots: Vec<Slot>;
+        {
+            let my_bank_forks = bank_forks.read().unwrap();
+            filtered_slots = result
+                .slots_to_repair
+                .into_iter()
+                .filter(|slot| {
+                    if slot <= &root_slot || is_full_slots.contains(slot) {
+                        return false;
+                    }
+                    let is_full = my_bank_forks
+                        .get(*slot)
+                        .map_or(false, |bank| bank.is_frozen());
+                    if is_full {
+                        is_full_slots.insert(*slot);
+                    }
+                    !is_full
+                })
+                .collect();
+        }
+        filtered_slots.sort();
+        info!(
+            "Active peers: {} Slots to repair: {:?}",
+            result.active_percentage, &filtered_slots
+        );
+        if filtered_slots.is_empty() && result.active_percentage > RESTART_SUPERMAJORITY_THRESHOLD {
+            *slots_to_repair_for_wen_restart.write().unwrap() = vec![];
+            break;
+        }
+        {
+            *slots_to_repair_for_wen_restart.write().unwrap() = filtered_slots;
+        }
+        write_wen_restart_records(wen_restart_path, progress)?;
+        let elapsed = timestamp().saturating_sub(start);
+        let time_left = GOSSIP_SLEEP_MILLIS.saturating_sub(elapsed);
+        if time_left > 0 {
+            sleep(Duration::from_millis(time_left));
+        }
+    }
+    progress.set_state(RestartState::HeaviestFork);
+    Ok(())
+}
+
+pub fn wait_for_wen_restart(
+    wen_restart_path: &PathBuf,
+    last_vote: VoteTransaction,
+    blockstore: Arc<Blockstore>,
+    cluster_info: Arc<ClusterInfo>,
+    bank_forks: Arc<RwLock<BankForks>>,
+    slots_to_repair_for_wen_restart: Option<Arc<RwLock<Vec<Slot>>>>,
+) -> Result<(), Box<dyn std::error::Error>> {
+    let mut progress = read_wen_restart_records(wen_restart_path)?;
+    while progress.state() != RestartState::Done {
```

Comment:
does this need to be a loop here? If I understand correctly you can never go backwards from `LastVotedForkSlots` to `Init` right?

Reply:
That is true, but we can restart from any state saved in local file.

Reply:
right, but we'll still only succeed each step once and move to the next step, the loop indicates differently to the reader. Something to indicate the linear fashion like:
```rust
if progress.state() == RestartState::Init {
  send_restart_last_voted_fork_slots(...)?
  progress.set_state(RestartState::LastVotedForkSlots);
  write_wen_restart_records(...)?;
}
if progress.state() == RestartState::LastVotedSlots {
  aggregate_restart_last_voted_fork_slots(...)?
  progress.set_state(RestartState::...);
  write_wen_restart_records(...)?;
}
...
if progress.state() != RestartState::Done {
  panic!();
}
```
could have a `progress.increment_state()` which moves it to the next step and writes the proto

Reply:
Changed.